File: Python/Uni_Grade.py
official_courses = {
"Math", "Physics", "Computer Science", "Biology", "Chemistry",
"Statistics", "English", "Economics", "History", "Philosophy",
"Sociology", "Political Science", "Geography", "Psychology",
"Art", "Music", "Engineering", "Law", "Medicine", "Business"
}

students = {}

def create_student():
	username = input("Enter a unique username: ")

	if username in students:
		print("Username already exists!")
		return

	name = input("Enter student name: ")
	age = int(input("Enter student age: "))
	city = input("Enter city: ")
	zip_code = input("Enter zip code: ")
    

	print("\nAvailable Courses:")
	print(", ".join(official_courses))
    
	courses = input("Enter courses (comma separated): ").split(",")
    
    # Clean spaces and only accept valid courses
	course_list = []
	for c in courses:
		c = c.strip()        # remove spaces
		if c in official_courses:   # only allow valid course
			if c not in course_list:  # avoid duplicates
				course_list.append(c)
    
    # Save student details inside dictionary
	students[username] = {
		"name": name,
		"age": age,
		"courses": course_list,    # use list for courses (simple)
		"address": {"city": city, "zip": zip_code}   # another dictionary for address
}
    
print("Student created successfully!\n")


def display_student(username):
    # Show full record
	if username in students:
		print(students[username])
	else:
		print("Student not found.")


def display_courses(username):
    # Show only courses
	if username in students:
		print("Courses:", students[username]["courses"])
	else:
		print("Student not found.")

def display_zip(username):
    # Show only zip
	if username in students:
		print("Zip:", students[username]["address"]["zip"])
	else:
		print("Student not found.")


def display_city(username):
    # Show only city
	if username in students:
		print("City:", students[username]["address"]["city"])
	else:
		print("Student not found.")


def add_course(username):
    # Add a new course
	if username not in students:
		print("Student not found.")
		return
    
	course = input("Enter course to add: ").strip()
    
	if course in official_courses:  # only allow valid course
		if course not in students[username]["courses"]:
			students[username]["courses"].append(course)
			print(f"{course} added successfully.")
		else:
			print("Course already enrolled.")
	else:
			print("Course not offered by department.")


def remove_course(username):
    # Remove a course
	if username not in students:
		print("Student not found.")
		return
    
	course = input("Enter course to remove: ").strip()
    
	if course in students[username]["courses"]:
		students[username]["courses"].remove(course)
		print(f"{course} removed.")
	else:
		print("Course not found in student’s list.")


def update_student(username):
    # Update student details
	if username not in students:
		print("Student not found.")
		return
    
	name = input("Enter new name (leave blank to skip): ")
	age = input("Enter new age (leave blank to skip): ")
	city = input("Enter new city (leave blank to skip): ")
	zip_code = input("Enter new zip (leave blank to skip): ")

	if name != "":
        	students[username]["name"] = name
	if age != "":
		students[username]["age"] = int(age)
	if city != "":
		students[username]["address"]["city"] = city
	if zip_code != "":
		students[username]["address"]["zip"] = zip_code
    
	print("Student record updated.")

def total_students():
    # Count number of students
	print(f"Total students: {len(students)}")

def menu():
	while True:
		print("\n===== Student Management System =====")
		print("1. Create Student")
		print("2. Display Student Record")
		print("3. Display Courses")
		print("4. Display Zip Code")
		print("5. Display City")
		print("6. Add Course")
		print("7. Remove Course")
		print("8. Update Student")
		print("9. Show Total Students")
		print("0. Exit")

		option = input("Choose an option: ")

		if option == "1":
			create_student()
		elif option == "2":
			username = input("Enter username: ")
			display_student(username)
		elif option == "3":
			username = input("Enter username: ")
			display_courses(username)
		elif option == "4":
			username = input("Enter username: ")
			display_zip(username)
		elif option == "5":
			username = input("Enter username: ")
			display_city(username)
		elif option == "6":
			username = input("Enter username: ")
			add_course(username)
		elif option == "7":
			username = input("Enter username: ")
			remove_course(username)
		elif option == "8":
			username = input("Enter username: ")
			update_student(username)
		elif option == "9":
			total_students()
		elif option == "0":
			print("Exiting... Goodbye!")
			break
		else:
			print("invalid")
	

menu()

File: JavaScript/UniGrade.js
const prompt = require('prompt-sync')();
let students = [];

// List of official courses
const officialCourses = [
  "Math", "Physics", "Computer Science", "Biology", "Chemistry",
  "English", "Economics", "History", "Law", "Medicine", "Business"
];

// Create a new student
function createStudent(username, name, age, city, zip, courses) {
  // Check if username already exists
  for (let std of students) {
    if (std.username === username) {
      console.log("Username already exists!");
      return;
    }
  }

    let courseList = [];
  for (let c of courseList) {
    if (officialCourses.includes(c) && !courseList.includes(c)) {
      courseList.push(c);
    }
  }

  // Make a new student object
  let student = {
    username: username,
    name: name,
    age: age,
    city: city,
    zip: zip,
    courses: courseList
  };

  // Add to list
  students.push(student);
  console.log("Student created successfully!");
}

// Display a student’s full record
function displayStudent(username) {
  for (let pupil of students) {
    if (pupil.username === username) {
      console.log(pupil);
      return;
    }
  }
  console.log("Student not found.");
}

// Display only courses
function displayCourses(username) {
  for (let pupil of students) {
    if (pupil.username === username) {
      console.log("Courses:", pupil.courses);
      return;
    }
  }
  console.log("Student not found.");
}

// Display city
function displayCity(username) {
  for (let pupil of students) {
    if (pupil.username === username) {
      console.log("City:", pupil.city);
      return;
    }
  }
  console.log("Student not found.");
}

// Display zip code
function displayZip(username) {
  for (let pupil of students) {
    if (pupil.username === username) {
      console.log("Zip:", s.zip);
      return;
    }
  }
  console.log("Student not found.");
}

// Add a course
function addCourse(username, course) {
  for (let pupil of students) {
    if (pupil.username === username) {
      if (!officialCourses.includes(course)) {
        console.log("Course not offered.");
        return;
      }
      if (pupil.courses.includes(course)) {
        console.log("Already enrolled.");
      } else {
        pupil.courses.push(course);
        //console.log(${course} added.);
      }
      return;
    }
  }
  console.log("Student not found.");
}

// Remove a course
function removeCourse(username, course) {
  for (let pupil of students) {
    if (pupil.username === username) {
      let index = pupil.courses.indexOf(course);
      if (index !== -1) {
        pupil.courses.splice(index, 1);
       // console.log(${course} removed.);
      } else {
        console.log("Course not found.");
      }
      return;
    }
  }
  console.log("Student not found.");
}

// Update student details
function updateStudent(username, newName, newAge, newCity, newZip) {
  for (let pupil of students) {
    if (pupil.username === username) {
      if (newName) pupil.name = newName;
      if (newAge) pupil.age = newAge;
      if (newCity) pupil.city = newCity;
      if (newZip) pupil.zip = newZip;
      console.log("Student updated.");
      return;
    }
  }
  console.log("Student not found.");
}

// Show total students
function totalStudents() {
  console.log("Total students:", students.length);
}


function menu() {
      while (true) {
        let choice = prompt(
          "===== Student Management System =====\n" +
          "1. Create Student\n" +
          "2. Display Student\n" +
          "3. Display Courses\n" +
          "4. Add Course\n" +
          "5. Remove Course\n" +
          "6. Update Student\n" +
          "7. Show Total Students\n" +
          "0. Exit\n\n" +
          "Enter choice:"
        );

        if (choice === "1") createStudent();
        else if (choice === "2") displayStudent();
        else if (choice === "3") displayCourses();
        else if (choice === "4") addCourse();
        else if (choice === "5") removeCourse();
        else if (choice === "6") updateStudent();
        else if (choice === "7") totalStudents();
        else if (choice === "0") {
          console.log("Exiting...");
          break;
        } else {
          console.log("Invalid choice, try again.");
        }
      }
    }

  menu()